tower/src/util/mod.rs

```diff
@@ -215,6 +220,216 @@ pub trait ServiceExt<Request>: tower_service::Service<Request> {
         MapErr::new(self, f)
     }
 
+    /// Composes a function after the service, regardless of whether the future
+    /// succeeds or fails.
+    ///
+    /// This is similar to the [`map_response`] and [`map_err] combinators,
+    /// except that the *same* function is invoked when the service's future
+    /// completes, whether it completes successfully or fails. This function
+    /// takes the `Result` returned by the service's future, and returns a
+    /// `Result`.
+    ///
+    /// Like the standard library's [`Result::and_then`], this method can be
+    /// used to implement control flow based on `Result` values. For example, it
+    /// may be used to  implement error recovery, by turning some `Err`
+    /// responses from the service into `Ok` responses. Similarly, some
+    /// successful responses from the service could be rejected, by returning an
+    /// `Err` conditionally, depending on the value inside the `Ok`. Finally,
+    /// this method can also be used to implement behaviors that must run when a
+    /// service's future completes, regardless of whether it succeeded or failed.
+    ///
+    /// This method can be used to change the [`Response`] type of the service
+    /// into a different type. It can also be used to change the [`Error`] type
+    /// of the service. However, because the `map_result` function is not applied
+    /// to the errors returned by the service's [`poll_ready`] method, it must
+    /// be possible to convert the service's [`Error`] type into the error type
+    /// returned by the `map_result` function. This is trivial when the function
+    /// returns the same error type as the service, but in other cases, it can
+    /// be useful to use [`BoxError`] to erase differing error types.
```

Comment:
Is box error a type we have now?

Reply:
https://github.com/tower-rs/tower/blob/1a845433177b31ebc6daff765ee81468c42d6676/tower/src/lib.rs#L85-L86